SmartDeviceLink/SDLHapticManager.m

```diff
@@ -0,0 +1,161 @@
+        [self.focusableViews addObject:currentView];
+        return;
+    }
+    // if current view has focusable sub views parse them recursively
+    else if (currentView.subviews.count > 0) {
+        NSArray<UIView *> *subviews = currentView.subviews;
+        
+        for (UIView *childView in subviews) {
+            [self sdl_parseViewHierarchy:childView];
+        }
+    }
+    //else just return
+    else {
+        return;
+    }
+}
+
+/**
+ Iterates through the focusable views, extracts rectangular parameters, creates Haptic RPC request and sends it
+ */
+- (void)sdl_sendHapticRPC {
+    NSMutableArray<SDLHapticRect *> *hapticRects = [[NSMutableArray alloc] init];
+    
+    for (UIView *view in self.focusableViews) {
+        CGPoint originOnScreen = [view.superview convertPoint:view.frame.origin toView:nil];
+        CGRect convertedRect = {originOnScreen, view.bounds.size};
+        SDLRectangle* rect = [[SDLRectangle alloc] initWithCGRect:(convertedRect)];
+        // using the view index as the id field in SendHapticData request (should be guaranteed unique)
+        NSUInteger rectId = [self.focusableViews indexOfObject:view];
+        SDLHapticRect *hapticRect = [[SDLHapticRect alloc] initWithId:(UInt32)rectId rect:rect];
+        [hapticRects addObject:hapticRect];
+    }
+    
+    SDLSendHapticData* hapticRPC = [[SDLSendHapticData alloc] initWithHapticRectData:hapticRects];
+    [self.connectionManager sendManagerRequest:hapticRPC withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+    }];
+}
+
+#pragma mark SDLHapticHitTester functions
+- (nullable UIView *)viewForSDLTouch:(SDLTouch *)touch {
+    UIView *selectedView = nil;
+    
+    for (UIView *view in self.focusableViews) {
+        //Convert the absolute location to local location and check if that falls within view boundary
+        CGPoint localPoint = [view convertPoint:touch.location fromView:self.projectionWindow];
+        if ([view pointInside:localPoint withEvent:nil]) {
+            if (selectedView != nil) {
+                selectedView = nil;
+                break;
+                //the point has been indentified in two views. We cannot identify which with confidence.
+            } else {
+                selectedView = view;
+            }
+        }
+    }
+    
+    if(selectedView != nil) {
+        NSUInteger selectedViewIndex = [self.focusableViews indexOfObject:selectedView];
+        
+        if (selectedViewIndex != NSNotFound) {
+            return self.focusableViews[selectedViewIndex];
+        }
+    }
+    return nil;
+}
+
+#pragma mark notifications
+/**
+ Function that gets called when projection view updated notification occurs.
+
+ @param notification object with notification data
+ */
+- (void)sdl_projectionViewUpdated:(NSNotification *)notification {
+    [self updateInterfaceLayout];
```

Comment:
It does not appear to be tested that a notification send causes an update of the interface layout.

Reply:
We added a test to cover this case as well.